### PR TITLE
Retry updating the EdgeDevice status

### PR DESCRIPTION
**Pull request by masayag**
The PR uses the patch method for updating the edge-device status.
In addition, it retries 3 times in case of a failure when the device
posts its information to the operator.

ECOPROJECT-255

Signed-off-by: Moti Asayag <masayag@redhat.com>

### CODE REVIEW

File: internal/yggdrasil/yggdrasil.go [hunk 1]
```diff
@@ -234,6 +239,29 @@ func (h *Handler) PostDataMessageForDevice(ctx context.Context, params yggdrasil
...
+func (h *Handler) updateDeviceStatus(ctx context.Context, device *v1alpha1.EdgeDevice, updateFunc func(d *v1alpha1.EdgeDevice)) middleware.Responder {
+	patch := client.MergeFrom(device.DeepCopy())
+	updateFunc(device)
+	err := h.deviceRepository.PatchStatus(ctx, device, &patch)
+	if err == nil {
+		return nil
+	}
+
+	// retry patching the edge device status
+	for i := 0; i < 3; i++ {
```
Comment (jakub-dzon) [inline internal/yggdrasil/yggdrasil.go:251]: What do you think about pausing between retries? 
Exponential back-off might be helpful if there is a lot of changes happening to the device at the same time.

File: internal/yggdrasil/yggdrasil.go [hunk 2]
```diff
@@ -168,7 +169,10 @@ func (h *Handler) PostDataMessageForDevice(ctx context.Context, params yggdrasil
 	case "heartbeat":
 		heartbeat := models.Heartbeat{}
 		contentJson, _ := json.Marshal(msg.Content)
-		json.Unmarshal(contentJson, &heartbeat)
+		err := json.Unmarshal(contentJson, &heartbeat)
+		if err != nil {
+			return nil
```
Comment (jakub-dzon) [inline internal/yggdrasil/yggdrasil.go:174]: That should be an error (HTTP 400)  instead of `nil`

File: internal/yggdrasil/yggdrasil.go [hunk 3]
```diff
@@ -195,7 +200,10 @@ func (h *Handler) PostDataMessageForDevice(ctx context.Context, params yggdrasil
 			if errors.IsNotFound(err) {
 				contentJson, _ := json.Marshal(msg.Content)
 				registrationInfo := models.RegistrationInfo{}
-				json.Unmarshal(contentJson, &registrationInfo)
+				err := json.Unmarshal(contentJson, &registrationInfo)
+				if err != nil {
+					return nil
```
Comment (jakub-dzon) [inline internal/yggdrasil/yggdrasil.go:205]: That also should be HTTP 400 instead of `nil`

File: internal/yggdrasil/yggdrasil.go [hunk 4]
```diff
@@ -234,6 +239,29 @@ func (h *Handler) PostDataMessageForDevice(ctx context.Context, params yggdrasil
...
+	updateFunc(device)
+	err := h.deviceRepository.PatchStatus(ctx, device, &patch)
+	if err == nil {
+		return nil
+	}
+
+	// retry patching the edge device status
+	for i := 0; i < 3; i++ {
+		device2, err := h.deviceRepository.Read(ctx, device.Name, device.Namespace)
+		if err == nil {
```
Comment (jakub-dzon) [inline internal/yggdrasil/yggdrasil.go:253]: What do you think about inverting the condition to eliminate nesting?

File: controllers/edgedeployment_controller.go [hunk 5]
```diff
@@ -184,8 +185,9 @@ func (r *EdgeDeploymentReconciler) addDeploymentsToDevices(ctx context.Context,
 	for _, edgeDevice := range edgeDevices {
 		if !hasDeployment(edgeDevice, name) {
 			deploymentStatus := managementv1alpha1.Deployment{Name: name, Phase: managementv1alpha1.Deploying}
+			patch := client.MergeFrom(edgeDevice.DeepCopy())
 			edgeDevice.Status.Deployments = append(edgeDevice.Status.Deployments, deploymentStatus)
-			err := r.EdgeDeviceRepository.UpdateStatus(ctx, &edgeDevice)
```
Comment (jakub-dzon) [inline controllers/edgedeployment_controller.go:188]: `UpdateStatus` can be removed from the `EdgeDeviceRepository`; let's not leave too many options for updating status.

**Review by jakub-dzon — CHANGES_REQUESTED** [review]

File: internal/yggdrasil/yggdrasil.go (hunk 3 above)
Comment (masayag) [reply internal/yggdrasil/yggdrasil.go]: indeed. for sure not nil. I added 400 to swagger.

**Review by pkliczewski — COMMENTED** [review]
/lgtm

**Review by pkliczewski — APPROVED** [review]

File: internal/yggdrasil/yggdrasil.go [hunk 6]
```diff
@@ -177,63 +181,92 @@ func (h *Handler) PostDataMessageForDevice(ctx context.Context, params yggdrasil
...
-				device.Namespace = h.initialNamespace
-				device.Spec.OsImageId = registrationInfo.OsImageID
-				device.Finalizers = []string{YggdrasilConnectionFinalizer, YggdrasilWorkloadFinalizer}
-				err := h.deviceRepository.Create(ctx, &device)
-				if err != nil {
-					logger.Error(err, "Cannot save EdgeDevice")
-					return operations.NewPostDataMessageForDeviceInternalServerError()
-				}
-
+			if !errors.IsNotFound(err) {
```
Comment (jakub-dzon) [inline internal/yggdrasil/yggdrasil.go:200]: :+1:

**Review by jakub-dzon — APPROVED** [review]

**Pull request merged**